### PR TITLE
Now we can override attributes with methods, refs #10134

### LINKED ISSUE #10134: INTERNAL ERROR with `__hash__ = None`
**Crash Report**

Mypy crashed when run on the following reduced file. Setting `__hash__` to `None` is the way to suppress Python's default hash method.

**Traceback**

```
> mypy mypy_crash.py --show-traceback
mypy_crash.py:2: error: Incompatible types in assignment (expression has type "None", base class "object" defined the type as "Callable[[object], int]")
mypy_crash.py:5: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.820+dev.15bd486a67a3214fdb854ba2f1d6d524113da6c1
Traceback (most recent call last):
  File "/home/user/.virtualenvs/mypy/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/__main__.py", line 11, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/main.py", line 98, in main
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 180, in build
    sources, options, alt_lib_path, flush_errors, fscache, stdout, stderr, extra_plugins
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 253, in _build
    graph = dispatch(sources, manager, stdout)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 2688, in dispatch
    process_graph(graph, manager)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 3012, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 3119, in process_stale_scc
    if not graph[id].type_check_second_pass():
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/build.py", line 2181, in type_check_second_pass
    return self.type_checker().check_second_pass()
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 344, in check_second_pass
    self.check_partial(node)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 356, in check_partial
    self.accept(node)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 401, in accept
    stmt.accept(self)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/nodes.py", line 687, in accept
    return visitor.visit_func_def(self)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 726, in visit_func_def
    self._visit_func_def(defn)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 737, in _visit_func_def
    self.check_method_override(defn)
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 1420, in check_method_override
    if self.check_method_or_accessor_override_for_base(defn, base):
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 1448, in check_method_or_accessor_override_for_base
    if self.check_method_override_for_base_with_name(defn, name, base):
  File "/home/user/.virtualenvs/mypy/lib/python3.6/site-packages/mypy/checker.py", line 1509, in check_method_override_for_base_with_name
    assert False, str(base_attr.node)
AssertionError: Var(__hash__)
mypy_crash.py:5: : note: use --pdb to drop into pdb
```

**To Reproduce**

Contents of mypy_crash.py:
```python
class Base:
    __hash__ = None

class Dervied(Base):
    def __hash__(self) -> int:
        return 0
```

**Your Environment**

- Mypy version used: master (0.820+dev.15bd486a67a3214fdb854ba2f1d6d524113da6c1)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.6.12 (compiled from source with pyenv)
- Operating system and version: Ubuntu 20.04

### LINKED ISSUE #11556: Complex inheritance generates INTERNAL ERROR
**Crash Report**

On HomeAssistant project, a complex flow is generating an INTERNAL ERROR when trying to add type hints.
https://github.com/home-assistant/core/pull/59503
I was able to reproduce the issue with a trimmed version of the code (see below).

**Traceback**

```
internal_error.py:36: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.920+dev.053a1beb94ee4e5b3260725594315d1b6776e42f
Traceback (most recent call last):
  File "/usr/local/bin/mypy", line 8, in <module>
    sys.exit(console_entry())
  File "/usr/local/lib/python3.8/dist-packages/mypy/__main__.py", line 11, in console_entry
    main(None, sys.stdout, sys.stderr)
  File "/usr/local/lib/python3.8/dist-packages/mypy/main.py", line 92, in main
    res, messages, blockers = run_build(sources, options, fscache, t0, stdout, stderr)
  File "/usr/local/lib/python3.8/dist-packages/mypy/main.py", line 170, in run_build
    res = build.build(sources, options, None, flush_errors, fscache, stdout, stderr)
  File "/usr/local/lib/python3.8/dist-packages/mypy/build.py", line 180, in build
    result = _build(
  File "/usr/local/lib/python3.8/dist-packages/mypy/build.py", line 256, in _build
    graph = dispatch(sources, manager, stdout)
  File "/usr/local/lib/python3.8/dist-packages/mypy/build.py", line 2712, in dispatch
    process_graph(graph, manager)
  File "/usr/local/lib/python3.8/dist-packages/mypy/build.py", line 3043, in process_graph
    process_stale_scc(graph, scc, manager)
  File "/usr/local/lib/python3.8/dist-packages/mypy/build.py", line 3150, in process_stale_scc
    if not graph[id].type_check_second_pass():
  File "/usr/local/lib/python3.8/dist-packages/mypy/build.py", line 2197, in type_check_second_pass
    return self.type_checker().check_second_pass()
  File "/usr/local/lib/python3.8/dist-packages/mypy/checker.py", line 366, in check_second_pass
    self.check_partial(node)
  File "/usr/local/lib/python3.8/dist-packages/mypy/checker.py", line 377, in check_partial
    self.accept(node)
  File "/usr/local/lib/python3.8/dist-packages/mypy/checker.py", line 422, in accept
    stmt.accept(self)
  File "/usr/local/lib/python3.8/dist-packages/mypy/nodes.py", line 727, in accept
    return visitor.visit_func_def(self)
  File "/usr/local/lib/python3.8/dist-packages/mypy/checker.py", line 754, in visit_func_def
    self._visit_func_def(defn)
  File "/usr/local/lib/python3.8/dist-packages/mypy/checker.py", line 765, in _visit_func_def
    self.check_method_override(defn)
  File "/usr/local/lib/python3.8/dist-packages/mypy/checker.py", line 1451, in check_method_override
    if self.check_method_or_accessor_override_for_base(defn, base):
  File "/usr/local/lib/python3.8/dist-packages/mypy/checker.py", line 1479, in check_method_or_accessor_override_for_base
    if self.check_method_override_for_base_with_name(defn, name, base):
  File "/usr/local/lib/python3.8/dist-packages/mypy/checker.py", line 1540, in check_method_override_for_base_with_name
    assert False, str(base_attr.node)
AssertionError: Var(async_specific)
internal_error.py:36: : note: use --pdb to drop into pdb
```

**To Reproduce**

```python
from typing import Any, Dict, TypedDict, cast

GenericServiceInfoType = Dict[str, Any]


class SpecificServiceInfo(TypedDict):
    host: str
    port: int


class BaseFlow:
    async def async_generic(
        self,
        service_info: GenericServiceInfoType,
    ) -> bool:
        return True

    async def async_specific(
        self,
        service_info: SpecificServiceInfo,
    ) -> bool:
        return await self.async_generic(cast(dict, service_info))


class StandardFlow(BaseFlow):
    async def async_standard_generic(
        self,
        service_info: GenericServiceInfoType,
    ) -> bool:
        return True

    async_specific = async_standard_generic  # type: ignore


class ExtendedFlow(StandardFlow):
    async def async_specific(
        self,
        service_info: SpecificServiceInfo,
    ) -> bool:
        return await self.async_generic(cast(dict, service_info))
```

**Your Environment**

- Mypy version used: 0.920+dev.053a1beb94ee4e5b3260725594315d1b6776e42f
- Mypy command-line flags: 
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.8.10
- Operating system and version: WSL (Ubuntu-20.04)

### PR DESCRIPTION
**Pull request by sobolevn**
It was crashing before this patch.

Closes #10134
Closes #11556
Refs #5425

### CODE REVIEW

File: test-data/unit/check-classes.test [hunk 1]
```diff
@@ -114,6 +114,43 @@ class A:
...
 
+[case testOverrideAttributeWithMethod]
+# This was crashing:
+# https://github.com/python/mypy/issues/10134
+from typing import Protocol
+
+class Base:
+    __hash__ = None
+
+class Dervied(Base):
```
Comment (AlexWaygood) [inline test-data/unit/check-classes.test:125]:
```suggestion
class Derived(Base):
```

File: test-data/unit/check-classes.test [hunk 2]
```diff
@@ -114,6 +114,43 @@ class A:
...
+
+[case testOverrideMethodWithAttribute]
+# The reserse should not crash as well:
+from typing import Callable
+
+class Base:
+    def __hash__(self) -> int:
+        pass
+
+class Dervied(Base):
```
Comment (AlexWaygood) [inline test-data/unit/check-classes.test:150]:
```suggestion
class Derived(Base):
```

File: test-data/unit/check-classes.test [hunk 3]
```diff
@@ -114,6 +114,43 @@ class A:
...
+
+class CorrectBase:
+    attr: CallableProtocol
+
+class CorrectDerived(CorrectBase):
+    def attr(self, arg: int) -> int:
+        pass
+
+[case testOverrideMethodWithAttribute]
+# The reserse should not crash as well:
```
Comment (AlexWaygood) [inline test-data/unit/check-classes.test:143]:
```suggestion
# The reverse should not crash as well:
```

**Review by AlexWaygood — COMMENTED** [review]
A few typos :)

**Comment by sobolevn** [comment]
@AlexWaygood thanks!

**Review by 97littleleaf11 — APPROVED** [review]
lgtm!

**Pull request merged**